nomos-da/network/core/src/protocols/dispersal/validator/behaviour.rs

```diff
@@ -145,3 +145,430 @@ impl<M: MembershipHandler<Id = PeerId, NetworkId = SubnetworkId> + 'static> Netw
+        }
+
+        // Collect blob message senders from executors
+        let mut message_senders: Vec<UnboundedSender<(u32, DaBlob)>> = Vec::new();
+
+        for i in 0..ALL_INSTANCES / GROUPS {
+            let blob_sender_0 = executor_0_swarms[i].behaviour().blobs_sender();
+            let blob_sender_1 = executor_1_swarms[i].behaviour().blobs_sender();
+            message_senders.extend(vec![blob_sender_0, blob_sender_1]);
+        }
+
+        let mut executor_tasks = vec![];
+        let (terminator_tx, terminator_rx) = watch::channel::<()>(());
+
+        // Spawn executors
+        for i in (0..ALL_INSTANCES / GROUPS).rev() {
+            let (terminator_0, terminator_1) = (terminator_rx.clone(), terminator_rx.clone());
+
+            let swarm = executor_0_swarms.remove(i);
+            let executor_0_poll =
+                async move { run_executor_swarm(swarm, MESSAGES_TO_SEND, terminator_0).await };
+
+            let swarm = executor_1_swarms.remove(i);
+            let executor_1_poll =
+                async move { run_executor_swarm(swarm, MESSAGES_TO_SEND, terminator_1).await };
+
+            executor_tasks.extend(vec![
+                tokio::spawn(executor_0_poll),
+                tokio::spawn(executor_1_poll),
+            ]);
+        }
+
+        // Send messages from all executors
+        for i in (0..ALL_INSTANCES / 2).rev() {
+            let sender = message_senders.remove(i);
+            let send_messages_task = async move {
+                send_dispersal_messages(sender, if i % 2 == 0 { 0 } else { 1 }, MESSAGES_TO_SEND)
+                    .await;
+            };
+            tokio::spawn(send_messages_task);
+        }
+
+        let mut validator_tasks = vec![];
+
+        // Spawn validators
+        for i in (0..ALL_INSTANCES / GROUPS).rev() {
+            let swarm = validator_0_swarms.remove(i);
+            let validator_0_poll = async { run_validator_swarm(swarm).await };
+
+            let swarm = validator_1_swarms.remove(i);
+            let validator_1_poll = async { run_validator_swarm(swarm).await };
+
+            validator_tasks.extend(vec![
+                tokio::spawn(validator_0_poll),
+                tokio::spawn(validator_1_poll),
+            ]);
+        }
+
+        let mut dispersal_request_counter = (0usize, 0usize);
+
+        for task in validator_tasks {
+            let requested = task.await.unwrap();
+            debug!(
+                "Validator received {:?} messages from subnet 0",
+                requested.0
+            );
+            debug!(
+                "Validator received {:?} messages from subnet 1\n",
+                requested.1
+            );
+            dispersal_request_counter = (
+                dispersal_request_counter.0 + requested.0,
+                dispersal_request_counter.1 + requested.1,
+            );
+        }
+
+        // Terminate any remaining executors
+        terminator_tx.send(()).unwrap();
```

Comment:
@romanzac @danielSanchezQ @bacv 
In continuation to #902 , I am trying to identify what could be wrong with this PR. But the main thing is I can't do proper debugging as of now, so I can't exactly confirm it...

This statement is executed in the main thread which is wrong I guess because after spawning(it is async so it is not waiting) the `for loop` at line 528 it will directly come to this line..and it will tell all executors
to drop off... That is why I guess there is error coming...

Please check into this one...
Also I am not good at devops so I can't really tell what is wrong with the pipeline tests😅...

Reply:
Let me add latest log from self-hosted 
[run3.log](https://github.com/user-attachments/files/17957722/run3.log)

Reply:
@bacv Would it be possible to reboot self-hosted machines just to run this workflow ? `Test Validator Behaviour` This run to check if state of the machine has any impact.

Reply:
Issues like 902 usually happen because multiple factors, e.g. environment is not friendly + connections are not that robust.

Reply:
I would suggest to:
1) Look at the runner machine. Configuration differences to macos-latest. State of CPU, memory, networking.
2) Play round libp2p networking layer. Use different transport protocol. Try any tuning params for QUICK.

Reply:
> This run to check if state of the machine has any impact.

This can be tested on gh linux runner, without requiring any changes to the self hosted. Let me double check the test, I feel that the state of a machine isn't significant factor in this case.

Reply:
I found only one log message type is present in self-hosted and not in macos-latest:

```
2024-11-29T10:04:19.560125Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor event: ConnectionClosed { peer_id: PeerId("12D3KooWE7XHQBnZ57Bk9tfpiTHN18cpV1B96tf4mi6RsdBRYtoJ"), connection_id: ConnectionId(21), endpoint: Dialer { address: /ip4/127.0.0.1/udp/5209/quic-v1/p2p/12D3KooWE7XHQBnZ57Bk9tfpiTHN18cpV1B96tf4mi6RsdBRYtoJ, role_override: Dialer }, num_established: 0, cause: Some(IO(Custom { kind: Other, error: Connection(ConnectionError(ApplicationClosed(ApplicationClose { error_code: 0, reason: b"" }))) })) }
```

Reply:
> nux runner,

Is self-hosted a linux ? We have already one differential run on macos-latest - np.